# Leçon 02C — Exercices résolus
*Structures linéaires -- Les dictionnaires --*

Notions utilisées : `Dico[c]`, `Dico[c] = e`, `get`, `pop`, `keys`, `values`, `items`, `len`, et le test `c in Dico` (utilisé dans l'exercice 1).

## Exercice 1

> Sans exécuter le code, indiquer ce qu'affiche chaque instruction :

### Réponses (à trouver de tête)

| Instruction | Affichage | Explication |
|:---|:---|:---|
| `print(notes["NSI"])` | `17` | valeur associée à la clé `"NSI"` |
| `print(len(notes))` | `3` | le dictionnaire contient 3 paires |
| `print("physique" in notes)` | `False` | `"physique"` n'est pas une clé |
| `print(notes.get("physique", "matière non suivie"))` | `matière non suivie` | clé absente : `get` renvoie la valeur par défaut donnée en 2ᵉ argument (sans elle, ce serait `None`) |
| `notes["anglais"] = 14` | *(rien)* | la clé existe : sa valeur est modifiée |
| `print(notes["anglais"])` | `14` | nouvelle valeur |

### Vérification (une fois qu'on a répondu !)

In [1]:
notes = {"maths": 15, "NSI": 17, "anglais": 12}
print(notes["NSI"])
print(len(notes))
print("physique" in notes)
print(notes.get("physique", "matière non suivie"))
notes["anglais"] = 14
print(notes["anglais"])

17
3
False
matière non suivie
14


## Exercice 2

> On considère le dictionnaire `capitales = {"France": "Paris", "Italie": "Rome", "Espagne": "Madrid"}`

In [2]:
capitales = {"France": "Paris", "Italie": "Rome", "Espagne": "Madrid"}

# 1) Ajouter : la clé n'existe pas, Dico[c] = e l'ajoute
capitales["Allemagne"] = "Berlin"
print(capitales)

# 2) Modifier : la clé existe, Dico[c] = e remplace la valeur
capitales["Espagne"] = "MADRID"
print(capitales)

# 3) Supprimer avec Dico.pop(c)
capitales.pop("Italie")
print(capitales)

# 4) Parcourir les paires (cle, valeur) avec items()
for pays, ville in capitales.items():
    print("La capitale de " + pays + " est " + ville + ".")

{'France': 'Paris', 'Italie': 'Rome', 'Espagne': 'Madrid', 'Allemagne': 'Berlin'}
{'France': 'Paris', 'Italie': 'Rome', 'Espagne': 'MADRID', 'Allemagne': 'Berlin'}
{'France': 'Paris', 'Espagne': 'MADRID', 'Allemagne': 'Berlin'}
La capitale de France est Paris.
La capitale de Espagne est MADRID.
La capitale de Allemagne est Berlin.


`Dico[c] = e` sert **à la fois** à ajouter une paire (clé nouvelle) et à modifier une valeur (clé existante). `items()` donne les paires `(cle, valeur)`, que l'on récupère dans deux variables `pays, ville`.

## Exercice 3

> Écrire une fonction **`age_de(prenom, annuaire)`** qui renvoie l'âge de la personne, ou **`-1`** si le prénom n'est pas dans l'annuaire.

In [3]:
def age_de(prenom, annuaire):
    # On vérifie d'abord que la clé existe (sinon annuaire[prenom] provoquerait une erreur)
    if prenom in annuaire:
        return annuaire[prenom]
    else:
        return -1

annuaire = {"Alice": 16, "Hugo": 17, "Inès": 15}
print(age_de("Hugo", annuaire))    # 17
print(age_de("Zoé", annuaire))     # -1

17
-1


**Variante :** `return annuaire.get(prenom, -1)` fait la même chose en une ligne. Comme dans l'exercice 1, `get` renvoie la valeur par défaut quand la clé est absente.

## Exercice 4

> Écrire une fonction **`occurrences(texte)`** qui renvoie un dictionnaire donnant le nombre d'apparitions de chaque caractère.

In [4]:
def occurrences(texte):
    compte = {}                        # dictionnaire vide
    for caractere in texte:            # on parcourt la chaîne caractère par caractère
        if caractere in compte:
            compte[caractere] = compte[caractere] + 1   # déjà vu : +1
        else:
            compte[caractere] = 1                       # première apparition
    return compte

print(occurrences("abracadabra"))

{'a': 5, 'b': 2, 'r': 2, 'c': 1, 'd': 1}


Résultat attendu : `{'a': 5, 'b': 2, 'r': 2, 'c': 1, 'd': 1}`

C'est un schéma à connaître : le **compteur avec dictionnaire**. Chaque caractère sert de clé et la valeur compte ses apparitions.

## Exercice 5

> `notes = {"maths": 15, "NSI": 17, "anglais": 12, "philo": 10}`

In [5]:
notes = {"maths": 15, "NSI": 17, "anglais": 12, "philo": 10}

# 1) Moyenne : somme des valeurs divisée par le nombre d'éléments
def moyenne(notes):
    somme = 0
    for note in notes.values():
        somme = somme + note
    return somme / len(notes)

# 2) Matière ayant la meilleure note
def meilleure(notes):
    matiere_max = None
    note_max = -1
    for matiere, note in notes.items():
        if note > note_max:
            note_max = note
            matiere_max = matiere
    return matiere_max

print(moyenne(notes))
print(meilleure(notes))

13.5
NSI


Résultat attendu : `13.5` puis `NSI`.

Pour la moyenne, seules les **valeurs** comptent, d'où `values()`. Pour la meilleure matière, il faut la note **et** la clé, d'où `items()`.

## Exercice 6

> Écrire une fonction **`inverse(d)`** qui renvoie un nouveau dictionnaire où les clés et les valeurs sont échangées.

In [6]:
def inverse(d):
    resultat = {}                   # nouveau dictionnaire : d n'est pas modifié
    for cle, valeur in d.items():
        resultat[valeur] = cle      # la valeur devient la clé, la clé devient la valeur
    return resultat

print(inverse({"a": 1, "b": 2, "c": 3}))

{1: 'a', 2: 'b', 3: 'c'}


Résultat attendu : `{1: 'a', 2: 'b', 3: 'c'}`

⚠️ Si deux clés ont la même valeur, une seule sera gardée, car une clé est unique dans un dictionnaire. Autre point : les clés doivent être des `int` ou des `str` (rappel du cours).